java/ffi/src/main/java/sleeper/foreign/FFIArray.java

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright 2022-2025 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.foreign;
+
+import jnr.ffi.NativeType;
+import jnr.ffi.Struct;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * Array class that must be used inside a JNR-FFI Struct. Creates a dynamic array that can be
+ * passed to C.
+ * Strong references are maintained for allocated memory so GC will dispose of
+ * memory when
+ * this object is collected.
+ *
+ * @param <T> object type of array
+ */
+public class FFIArray<T> {
+    /** Length of array. */
+    private final Struct.size_t len;
+
+    /** Pointer to base of dynamically allocated array. */
+    private final Struct.Pointer arrayBase;
+    /**
+     * Reference to dynamically allocated array to prevent GC until Array instance
+     * is collected.
+     */
+    private jnr.ffi.Pointer basePtr;
+    /**
+     * Reference to dynamically allocated items to prevent GC until Array instance
+     * is collected.
+     */
+    private jnr.ffi.Pointer[] items;
+
+    public FFIArray(Struct enclosing) {
+        this.len = enclosing.new size_t();
+        this.arrayBase = enclosing.new Pointer();
+    }
+
+    /**
+     * Array length.
+     *
+     * @return length of array
+     */
+    public long length() {
+        return len.get();
+    }
+
+    /**
+     * Create a dynamic array of items in this array.
```

Comment:
This should be descriptive, like "Creates".

See the Oracle style guide under "Use 3rd person (descriptive) not 2nd person (prescriptive)": https://www.oracle.com/uk/technical-resources/articles/java/javadoc-tool.html